File: Golang/22. Generate Parentheses.go
func generateParenthesis(n int) []string {
    ans := make([]string, 0, 2*n)
    recursion(&ans, "", n, 0, 0)
    return ans
}

func recursion(ans *[]string, s string, n, left, right int) {
    if len(s) == 2*n {
        *ans = append(*ans, s)
    }
    if left < n {
        recursion(ans, s+"(", n, left+1, right)
    }
    if right < left {
        recursion(ans, s+")", n, left, right+1)
    }
}

File: Python3/22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ans = []
        def recursion(s='', left=0, right=0):
            if len(s) == 2*n:
                ans.append(s)
            if left < n:
                recursion(s+'(', left+1, right)
            if left > right:
                recursion(s+')', left, right+1)
        recursion('')
        return ans
